functions/src/actions/wayback-machine.js

```diff
@@ -0,0 +1,125 @@
+
+// Local imports
+const config = require('../config');
+const {debug} = require('../utils/logger')('ia:actions:wayback-machine');
+const dialog = require('../dialog');
+const endpointProcessor = require('../network/endpoint-processor');
+const waybackStrings = require('../strings').intents.wayback;
+
+/**
+ * Handle wayback query action
+ * - fill slots of wayback query
+ * - perform data requests to archive and alexa rankings
+ * - construct response speech for action
+ *
+ * @param app
+ */
+function handler (app) {
+  // Create wayback object
+  const waybackObject = {
+    url: '',
+    earliestYear: 0,
+    latestYear: 0,
+    totalUniqueURLs: 0,
+    alexaWorldRank: 0,
+    alexaUSRank: 0,
+    speech: waybackStrings.default,
+  };
+
+  // Check parameter for Wayback qualifier
+  let wayback = app.params.getByName('wayback');
+  if (!wayback.includes('wayback')) {
+    debug('wayback action called by mistake');
+    dialog.ask(app, waybackObject);
+  }
+
+  // Get url parameter and make url queries
+  waybackObject.url = app.params.getByName('url');
+  const archiveQueryURL = endpointProcessor.preprocess(
+    config.wayback.ARCHIVE, app, waybackObject
+  );
+  const alexaQueryURL = endpointProcessor.preprocess(
+    config.wayback.ALEXA, app, waybackObject
+  );
+
+  return Promise.all([axios.get(archiveQueryURL), axios.get(alexaQueryURL)])
+    .then(function (allData) {
+      // All data available here in the order it was called.
+
+      // Parse data from archive request
+      let archiveJSON = allData[0].data;
+      archiveEngine(archiveJSON, waybackObject);
+
+      // Parse data from alexa request
+      let alexaJSON;
+      let XMLparser = new xml2js.Parser();
+      XMLparser.parseString(allData[1].data, function (err, result) {
+        if (err) {
+          debug('The XML parser didn\'t work');
+          waybackObject.speech = waybackStrings.error;
+          dialog.ask(app, waybackObject);
+        } else {
+          alexaJSON = JSON.parse(JSON.stringify(result));
+        }
+      });
+      alexaEngine(alexaJSON, waybackObject);
+
+      // Construct response dialog for action
+      if (waybackObject.alexaUSRank !== 0) {
+        waybackObject.speech = mustache.render(waybackStrings.speech, waybackObject);
+        waybackObject.speech += mustache.render(waybackStrings.additionalSpeech, waybackObject);
+      } else {
+        waybackObject.speech = mustache.render(waybackStrings.speech, waybackObject);
+        waybackObject.speech += '.';
+      }
+
+      dialog.close(app, waybackObject);
+    });
+} // End of handler
+
+function archiveEngine (archiveJSON, waybackObject) {
+  // Create array of capture years and then find earliest year
+  //  and most recent year.
+  let yearsArray = Object.keys(archiveJSON.captures);
+  waybackObject.earliestYear = yearsArray[0];
+  waybackObject.latestYear = yearsArray[yearsArray.length - 1];
+
+  // Traverse URL category
+  const traverse = obj => {
```

Comment:
pls put this function to `./utils` directory. Btw it is a good candidate for unit tests :)